proofcarryingdata/zupass
# Fix packaging and show example usage with different bundlers

Fixes #965 
Fixes #1136
Fixes #421 

Various third party/hackathon users reported issues when importing our packages.

Some major categories:
1) Packages include typescript source, and using them means compiling the typescript source, which requires having a bunch of type packages installed (e.g. https://github.com/proofcarryingdata/zupass/issues/1136)
2) Our builds only support certain package resolution strategies, which was an issue for the ZuAstro team as they were trying to use code from the Jubmoji team, which also only supported certain package resolution strategies, but not the same ones that we do
3) The passport-interface nightmare

In addition, using CommonJS rather than ESM by default causes some inflation of our bundle size, as esbuild can't do tree-shaking on CommonJS packages. The current version of this branch cuts `passport-client` bundle size from 7.7mb to 7.1mb as a result of using ESM packages wherever possible.

Before: 
<img width="458" alt="image" src="https://github.com/proofcarryingdata/zupass/assets/20022/ae5c49a5-d9d0-40d0-a12a-b0d32f1c0334">

After:
<img width="475" alt="image" src="https://github.com/proofcarryingdata/zupass/assets/20022/e2f64335-e2c6-4cdd-977c-e0e1c42c8c13">

In this branch, we have configuration to do both ESM and CommonJS builds of all packages, and some builds have both node and browser variants. Each package also generates a separate TypeScript definition file, rather than bundling the whole TypeScript source. This makes the build process take longer than it used to, though it also means that we can eliminate the separate type-checking phase of CI, since generating the definitions performs the same task and will flag up the same errors.

Most modules worked without any changes to the TypeScript source, or without any module-specific configuration, but there were a couple of exceptions in cases where we import third-party modules with some quirks. In these cases we can provide some configuration that saves the users of our packages from having to deal with these quirks themselves.

I've also added example apps for NextJS, Vite, Create-React-App (Webpack) and a server-side Express app that uses ESBuild. This shows that the packages work in a variety of different environments and configurations. There are probably still edge cases I haven't caught, but this covers all of the types of project I've seen people try to build.

This has been tested locally using a local NPM package mirror, with modified packages but no version number change. If we merge this then we should bump all of the package version numbers and update the examples to use these.

## Overview of changes

`apps/anon-message-client` has been changed to remove the hacks previously necessary for `passport-interface` to work.

`examples/*` includes various example apps. The overwhelming majority of the code in this PR is just the source for these examples.

`packages/*` - each package now has updated configuration for how it ought to be bundled, and a changed `build` command in `package.json`. `passport-ui` and `passport-crypto` have some other fixes that resolved some bundling issues, but no modules have large changes to their content. `styled-components` has some specific issues when bundled using ESM, which made it necessary to wrap it in `passport-ui` and then use this wrapper in the PCD packages, but this introduces no new dependencies for the affected packages since they depended on `passport-ui` anyway.

## How to test

Run `./test-packaing.sh`.

This will:
- Run Verdaccio locally using the config in `verdaccio.yml`
- Publish all `@pcd` packages to Verdaccio
- Reinstall each example app with Verdaccio as the registry, meaning that it will use the locally-published `@pcd` packages. Third-party packages will be proxied from `yarnkpkg` as per the Verdaccio config.
- Build each example app. The script will fail if any command fails.

This is a bit slow as it needs to download the packages each time, but it does give a faithful reproduction of the third-part developer experience.

Todo:
- [x] Try pushing these packages to npm under a different name, to see if they work properly when imported
- [x] Test with NextJS
- [x] Test with create-react-app
- [x] Test with vanilla Express
- [x] Test with Vite
